boards/BMS/Test/Src/Test_Eeprom.cpp

```diff
@@ -0,0 +1,207 @@
+#include "Test_Eeprom.h"
+
+extern "C"
+{
+#include "App_Eeprom.h"
+}
+
+#define MAX_FLOATS_PER_PAGE 4
+
+namespace EepromTest
+{
+FAKE_VALUE_FUNC(EEPROM_StatusTypeDef, read_page, uint16_t, uint8_t, uint8_t *, uint16_t);
+FAKE_VALUE_FUNC(EEPROM_StatusTypeDef, write_page, uint16_t, uint8_t, uint8_t *, uint16_t);
+FAKE_VALUE_FUNC(EEPROM_StatusTypeDef, page_erase, uint16_t);
+
+class BmsEepromTest : public testing::Test
+{
+  protected:
+    void SetUp() override
+    {
+        eeprom = App_Eeprom_Create(write_page, read_page, page_erase);
+
+        RESET_FAKE(read_page);
+        RESET_FAKE(write_page);
+        RESET_FAKE(page_erase);
+
+        read_page_fake.return_val  = EEPROM_OK;
+        write_page_fake.return_val = EEPROM_OK;
+        page_erase_fake.return_val = EEPROM_OK;
+    }
+
+    void TearDown() override { TearDownObject(eeprom, App_Eeprom_Destroy); }
+
+    struct Eeprom *eeprom;
+
+    // This function is a carbon copy of 'convert_bytes_to_float' located in App_Eeprom.c, however because it is static,
+    // in order to test it, it had to be copied here
+    static float convert_bytes_to_float(uint8_t *bytes_to_convert)
+    {
+        // Create union that stores float and byte array in same memory location.
+        // This allows you to access 8-bit segments of the float value using array indexing
+        union
+        {
+            float   float_val;
+            uint8_t bytes[4];
+        } u;
+
+        for (int i = 0; i < 4; i++)
+        {
+            u.bytes[i] = bytes_to_convert[i];
+        }
+
+        return u.float_val;
+    }
+};
+
+TEST_F(BmsEepromTest, test_float_converted_to_bytes_full_page)
+{
+    uint16_t page   = 1; // arbitrary value
+    uint8_t  offset = 0; // arbitrary value
+
+    uint8_t num_floats = MAX_FLOATS_PER_PAGE;
+
+    // Fill float array with random numbers so that they are not garbage values
+    float input_float_array[MAX_FLOATS_PER_PAGE] = { 1.0, 2.0, 3.0, 4.0 };
+
+    App_Eeprom_WriteFloats(eeprom, page, offset, input_float_array, num_floats);
+
+    // read_page_fake.arg2val holds the pointer to the float input converted to a byte array
+    uint8_t byte_values[PAGE_SIZE];
+
+    // capture converted byte array
+    for (int i = 0; i < PAGE_SIZE; i++)
+    {
+        byte_values[i] = write_page_fake.arg2_val[i];
+    }
+
+    // Unfortunately, cannot easily access the bytes converted back to floats due to FFF limitations
+    // Because function used is static and located in App_Eeprom.c, need to make a copy of that static function in this
+    // file This is NOT maintainable, any changes to convert_bytes_to_float must be added here as well.
```

Comment:
For this test, I wanted to write a floats and read them back to test that they are being converted to bytes correctly. With our current test/mock setup, there is no way for me to access the byte-array in App_Eeprom_ReadFloats. As such the next best thing is to test the function that converts bytes into floats directly, problem being that I have it implemented as a static function that is inaccessible to the test file. So the terrible bandaid solution was to copy the entire static function and have a copy of it word-for-word in the test file. This works and allows me to test it, but if there is ever a need to modify this function, it would need to be updated in the test file as well. 

Frankly, I don't know if there is a better way of doing this, besides making the conversation function no longer static, (I just don't know where I'd put it if I did that). I just wanted to write more tests, as I feel they are extremely valuable, but to say that I had to do some jerry-rigging to be able to write tests for these functions at all is an understatement.

tldr. static function not visible to test file. copied it here. it's ghetto, but at least it's tested.

Reply:
you guys should use some define magic to fix this:
```c
#if TEST
#define STATIC
#else
#define STATIC static
#endif // TEST
```

then define your function as:
```c
STATIC float convert_bytes_to_float...
```

Reply:
I tried something like this but ran into a couple issues (maybe you can clear some light on this)
1. even if I did this, wouldn't I still need to put function definitions in the header? I guess header definitions could be stuck within a #ifdef block

2. I'd have to figure out where to define TEST, as the test files are compiled last, so defining it at the start of a test before including App_Eeprom.h doesn't work. 

Another fix I saw was to include the .c file rather than the header in the test file, however this would re-compile anything included in the .c file, increasing executable size and compile time, but would include the static functions. This would require messing with CMake, as doing so right now results in the compiler not being able to find the .c file.

Reply:
1. yep, but that's fine. can add a separate section of function declarations with a comment
2. something like that may already be defined in the highest level CMakeLists.txt, or you can define it yourself

> Another fix I saw was to include the .c file rather than the header in the test file
seems really messy, the define magic is pretty industry standard afaik